### PR TITLE
Add the ability to have worker stick around between buck commands

### PR DESCRIPTION
**Pull request by mikekap**
This adds the idea of a `persistent` worker tool as mentioned in #712 . These workers have the same lifetime as `buckd`, unless the worker inputs themselves change, which causes a restart. We use the `RuleKey` via `InitializableFromDisk` to decide when that needs to happen.

### CODE REVIEW

File: docs/rule/worker_tool.soy [hunk 1]
```diff
@@ -75,6 +75,17 @@
...
+  {param name: 'persistent' /}
+  {param default: 'False' /}
+  {param desc}
+    If set to true, buck will not restart the tool unless the tool itself changes. This means the
+    tool will persist across multiple buck commands without being shut down and may see the same
+    rule being built more than once. Be careful when using this with tools that don't expect
+    to process the same input (with different contents) twice!
+  {/param}
+{/call}
+
```
Comment (Coneko) [inline docs/rule/worker_tool.soy]: I don't think this needs to be configurable at the tool level: a global configuration option in `.buckconfig` for migration purposes is ok, but generally worker tools should be safe to persist across multiple commands anyway.

From the tool's perspective there is no difference between being persisted or not.

File: src/com/facebook/buck/shell/ShBinary.java [hunk 2]
```diff
@@ -136,7 +136,8 @@ public Tool getExecutableCommand() {
   // for this rule to be usable.
   @Override
   public ImmutableSortedSet<BuildRule> getRuntimeDeps() {
-    return ImmutableSortedSet.copyOf(getResolver().filterBuildRuleInputs(main));
+    return ImmutableSortedSet.copyOf(getResolver().filterBuildRuleInputs(
```
Comment (Coneko) [inline src/com/facebook/buck/shell/ShBinary.java]: Nice catch!

Can you split this out in a separate PR?

File: src/com/facebook/buck/shell/WorkerProcess.java [hunk 3]
```diff
@@ -147,6 +149,26 @@ public void close() {
...
+          LOG.debug(e2, "Failed to read stderr file");
+        } finally {
+          try {
+            fileReader.get().close();
+          } catch (IOException e3) {
+            LOG.debug(e3, "Failed to close stderr reader");
+          }
+        }
+      }
+
```
Comment (Coneko) [inline src/com/facebook/buck/shell/WorkerProcess.java]: Please split out these changes in a separate PR.

File: src/com/facebook/buck/shell/WorkerProcessPool.java [hunk 4]
```diff
@@ -16,40 +16,36 @@
...
-import com.google.common.base.Optional;
+import com.google.common.hash.HashCode;
 
 import java.io.IOException;
 import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.Semaphore;
 
 public abstract class WorkerProcessPool {
 
-  public static final int UNLIMITED_CAPACITY = 0;
```
Comment (Coneko) [inline src/com/facebook/buck/shell/WorkerProcessPool.java]: Seems like a good change, but please pull it out into a separate PR.

File: src/com/facebook/buck/step/AbstractTestStep.java [hunk 5]
```diff
@@ -83,7 +83,7 @@ public StepExecutionResult execute(ExecutionContext context) throws InterruptedE
     }
     ProcessExecutorParams params = ProcessExecutorParams.builder()
         .setCommand(command)
-        .setDirectory(workingDirectory)
+        .setDirectory(workingDirectory.transform(filesystem::resolve))
```
Comment (Coneko) [inline src/com/facebook/buck/step/AbstractTestStep.java]: What is this change for?

**Comment by Coneko** [comment]
Seems generally ok, there's a lot of changes that should be split out into separate PRs.

Only practical problem I can see is that if the worker tool changes target we'll keep around persistent copies of all the different targets it had until the daemon restarts.

File: src/com/facebook/buck/shell/WorkerProcessProtocolZero.java [hunk 6]
```diff
@@ -239,6 +239,8 @@ public void close() throws IOException {
       processStdinWriter.close();
       processStdoutReader.endArray();
       processStdoutReader.close();
+    } catch (IllegalStateException ex) {
+      throw new IOException(ex);
```
Comment (mikekap) [inline src/com/facebook/buck/shell/WorkerProcessProtocolZero.java]: the `processStdoutReader.endArray()` can throw an `IllegalStateException`.

File: src/com/facebook/buck/step/AbstractTestStep.java (hunk 5 above)
Comment (Coneko) [inline src/com/facebook/buck/step/AbstractTestStep.java]: Make a separate PR just for this then.

File: src/com/facebook/buck/shell/WorkerProcessProtocolZero.java (hunk 6 above)
Comment (Coneko) [inline src/com/facebook/buck/shell/WorkerProcessProtocolZero.java]: But why do you need to wrap it in an `IOException`?

File: docs/rule/worker_tool.soy (hunk 1 above)
Comment (Coneko) [inline docs/rule/worker_tool.soy]: I'm just not sure we guarantee we will run the tools only once with certain inputs during the build to begin with, for example consider a multi platform build.

File: src/com/facebook/buck/shell/WorkerProcessProtocolZero.java (hunk 6 above)
Comment (mikekap) [inline src/com/facebook/buck/shell/WorkerProcessProtocolZero.java]: Opened #960 and reverted this. Originally the callers only caught `IOException` and this seemed IO-ish, but it's probably more error-proof to just catch `Exception` in callers.

File: docs/rule/worker_tool.soy (hunk 1 above)
Comment (mikekap) [inline docs/rule/worker_tool.soy]: Ah I see, I didn't realize that. I think from the perspective of the code, it's still one build though - each time a `BuildCommand` gets created a new `ExecutionContext` gets created with it (which is what holds the ephemeral workers right now).

The issue I'm facing is with the Clojure compiler (though I think Scala may have a similar issue). The basic problem is that loading files - compiled or not - is really slow because it executes a lot of `static`-level code. A good way to "cache" that load step is simply to not unload the output. The current `worker_tool` is perfect for that. Let's say I have 2 rules with dependencies: `//:a -> //:b`.
 - On a clean build of `//:a`, a new `worker_tool` simply compiles `//:b` and keeps it in-memory. Then when `//:a` is compiled, we add the output of `//:b` to the classpath, but the classloader ignores it since the compiled class is already loaded.
 - If we change `//:a`'s input and build `//:a` again, the new `worker_tool` will add `//:b`'s output to the classpath to load the compiled code rather than recompiling `//:b` again.
 - Unfortunately this doesn't work if a new "version" of `//:b` is compiled with the same worker - there's really no way to force unload a class in java cleanly.

With this approach the `worker_tool` never ends up reloading code unless it really needs to. The downside is that targets need precise deps or the builds won't be deterministic. The way that's accomplished in the project I linked above is via an autodeps-like mechanism that reads the code to figure out the exact dependency graph.

Is there a chance we can keep the two kinds of persistent processes coexisting (though perhaps not as a default)?

Comment (Coneko) [inline docs/rule/worker_tool.soy]: We can leave it a non default for now, but the exact semantics of what executions are going to be done in a single build seem very implementation specific, and not something you can rely on.

It really sounds like something that just happens to work right now and could break at any moment.

**Comment by Coneko** [comment]
Needs a rebase.

**Comment by Coneko** [comment]
Does the persistent worker die correctly if you `kill -9` `buckd`?

File: src/com/facebook/buck/shell/DefaultWorkerTool.java [hunk 7]
```diff
@@ -85,7 +103,47 @@ public int getMaxWorkers() {
...
+
+  @Override
+  public HashCode getInstanceKey() {
+    return buildOutputInitializer.getBuildOutput().getRuleKey().getHashCode();
+  }
+
+  @Override
+  public Data initializeFromDisk(OnDiskBuildInfo onDiskBuildInfo) throws IOException {
+    Optional<RuleKey> ruleKey =
+        onDiskBuildInfo.getRuleKey(BuildInfo.METADATA_KEY_FOR_RULE_KEY);
```
Comment (k21) [inline src/com/facebook/buck/shell/DefaultWorkerTool.java]: Why does the worker tool need to read its rule key from the disk? Cannot it use a `RuleKeyBuilderFactory`? I would prefer that over relying on mutable external state.

File: src/com/facebook/buck/rules/args/WorkerMacroArg.java [hunk 8]
```diff
@@ -99,6 +103,18 @@ public String getStartupArgs() {
     return workerTool.getArgs();
   }
 
+  public Optional<String> getPersistentWorkerKey() {
+    if (workerTool.isPersistent()) {
+      return Optional.of(buildTarget.toString());
```
Comment (k21) [inline src/com/facebook/buck/rules/args/WorkerMacroArg.java]: The key should probably include the target's cell in some form to avoid false sharing of workers between rules in different cells.

File: src/com/facebook/buck/shell/DefaultWorkerTool.java (hunk 7 above)
Comment (mikekap) [inline src/com/facebook/buck/shell/DefaultWorkerTool.java]: This seems to be a unified way to do something regardless whether a rule has been built or restored from cache (i.e. when steps aren't run). At the point `initializeFromDisk` is called, `OnDiskBuildInfo` actually has `getRuleKey` called on it quite a few times, so I think this is actually faster than rehashing the key via the factory. If not - seems like `OnDiskBuildInfo` could use a cache.

Comment (mikekap) [inline src/com/facebook/buck/shell/DefaultWorkerTool.java]: Sorry - I misread the question there. We could definitely recompute the rule key to avoid relying on the build info, but I don't know if that's much better. `RuleKeyBuilderFactory` will have to hash the source files anyway, which are likely build outputs and as mutable as the written rule key file. If anything, I think it would help if the data in `OnDiskBuildInfo` wasn't coming directly from disk but rather from the data with which the build was started.

**Comment by mikekap** [comment]
@Coneko not sure about the `kill -9` bit. I don't think it's possible to force terminate the worker processes, but they will definitely get their stdin closed. With some luck they will terminate too, but there's no way to be sure of it.

In case it's a more graceful exit, we do tear down the worker pools on daemon shutdown.

File: src/com/facebook/buck/rules/args/WorkerMacroArg.java (hunk 8 above)
Comment (k21) [inline src/com/facebook/buck/rules/args/WorkerMacroArg.java]: Unfortunately I think outside of the parser the cell prefix is not present in the build targets because it would have to change based on which cell references the target (two cells can be using a different names for another cell or the same name for different cells). Explicitly adding the `cellPath` from the build target should work well though.

File: src/com/facebook/buck/shell/DefaultWorkerTool.java (hunk 7 above)
Comment (k21) [inline src/com/facebook/buck/shell/DefaultWorkerTool.java]: The `RuleKeyBuilderFactory` classes have a cache of all computed rule keys on them, so it should be cheap. The part I worry about when using `OnDiskBuildInfo` is that it can only be safely called after the rule has been built and the call to it is buried in the call stack of a method that doesn't communicate this requirement in any way (`WorkerMacroArg#getWorkerHash`). But since violating that would result in an exception which should be caught by the tests, it is probably not such a big deal.

File: src/com/facebook/buck/rules/args/WorkerMacroArg.java (hunk 8 above)
Comment (mikekap) [inline src/com/facebook/buck/rules/args/WorkerMacroArg.java]: Ah I see. Makes sense - thanks for the explanation!

**Comment by Coneko** [comment]
Strange, github says it can be merged cleanly, but appveyor and our internal CI fail to patch it, can you rebase it please?
